calibration/recorder.py

```diff
@@ -0,0 +1,553 @@
+#!/usr/bin/env python
+
+"""Data recorder for Soil Power Sensor
+
+The recorder controls the SPS firmware and a Keithley 2400 Source
+Measurement Unit (SMU). Ideally the script should work with any microcontrollre
+flashed with the firmware and any source measurement unit that supports
+Standard Commands for Programable Instruments (SCPI). The units listed are the
+ones that the script was developed and tested on. It allows to step through
+a range of output voltages on the Keithley and measure the voltage and current
+from both the SMU and the Soil Power Sensor (SPS).
+
+Example
+-------
+The following examples shows how to collect data on voltage ranges 0-3.3V
+increasing the voltage by 0.1V every step. The SPS is connected to
+``/dev/ttyACM0`` and the SMu is connected to ``/dev/ttyUSB0``.::
+
+    $ python recorder.py 0 3.3 0.1 /dev/ttyACM0 /dev/ttyACM0
+
+To see a list of all CLI parameters:::
+
+    $ python recorder.py -h
+"""
+
+import pdb
+
+import time
+import argparse
+import socket
+import serial
+import numpy as np
+import pandas as pd
+import pdb
+from tqdm import tqdm
+from typing import Tuple
+from soil_power_sensor_protobuf import decode_measurement
+
+
+class SerialController:
+    """Generic serial controller that will open and close serial connections"""
+
+    # Serial port
+    ser = None
+
+    def __init__(self, port, baudrate=115200, xonxoff=False):
+        """Constructor
+
+        Initialises connection to serial port.
+
+        Parameters
+        ----------
+        port : str
+            Serial port of device
+        baudrate : int, optional
+            Baud rate for serial communication (default is 115200, STM32 functions at 115200)
+        xonxoff  : bool, optional
+            Flow control (default is on)
```

Comment:
Shoot when I fixed the software flow control I forgot to update this doc

Reply:
should default be off?